internal/pkg/agent/application/actions/handlers/handler_helpers.go

```diff
@@ -0,0 +1,174 @@
+				if unit.Type == client.UnitTypeInput && unit.Config != nil && unit.Config.Type == name {
+					ucs = append(ucs, unitWithComponent{unit, comp.Component})
+				}
+			}
+		}
+	}
+	return ucs
+}
+
+func contains[T comparable](arr []T, val T) bool {
+	for _, v := range arr {
+		if v == val {
+			return true
+		}
+	}
+	return false
+}
+
+type proxiedActionsNotifier struct {
+	log           *logp.Logger
+	performAction performActionFunc
+
+	timeout    time.Duration
+	minBackoff time.Duration
+	maxBackoff time.Duration
+}
+
+const (
+	defaultActionDispatcherTimeout    = 40 * time.Second
+	defaultActionDispatcherBackoffMin = 500 * time.Millisecond
+	defaultActionDispatcherBackoffMax = 10 * time.Second
+)
+
+func newProxiedActionsNotifier(log *logp.Logger, performAction performActionFunc) proxiedActionsNotifier {
+	return proxiedActionsNotifier{
+		log:           log,
+		performAction: performAction,
+		timeout:       defaultActionDispatcherTimeout,
+		minBackoff:    defaultActionDispatcherBackoffMin,
+		maxBackoff:    defaultActionDispatcherBackoffMax,
+	}
+}
+
+func (d proxiedActionsNotifier) notify(ctx context.Context, action dispatchableAction, ucs []unitWithComponent) error {
+	if action == nil {
+		return nil
+	}
+
+	// Deserialize the action into map[string]interface{} for dispatching over to the apps
+	params, err := action.MarshalMap()
+	if err != nil {
+		return err
+	}
+
+	actionType := action.Type()
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	dispatch := func(uc unitWithComponent) error {
+		if uc.unit.Config == nil {
+			return nil
+		}
+		d.log.Debugf("Dispatch %v action to %v", actionType, uc.unit.Config.Type)
+		res, err := d.performAction(ctx, uc.component, uc.unit, uc.unit.Config.Type, params)
+		if err != nil {
+			d.log.Debugf("%v failed to dispatch to %v, err: %v", actionType, uc.component.ID, err)
+			// ErrNoUnit means that the unit is not longer available
+			// This can happen if the policy change updated state while the action proxying was retried
+			// Stop retrying proxying action to that unit return nil
+			if errors.Is(err, runtime.ErrNoUnit) {
+				d.log.Debugf("%v unit is not longer managed by runtime, possibly due to policy change", uc.component.ID)
+				return nil
+			}
+			return err
+		}
+
+		strErr := readMapString(res, "error", "")
+		if strErr != "" {
+			d.log.Debugf("%v failed for %v, err: %v", actionType, uc.component.ID, strErr)
+			return errors.New(strErr)
+		}
+		return nil
+	}
+
+	dispatchWithBackoff := func(uc unitWithComponent) error {
+		ctx, cn := context.WithTimeout(ctx, d.timeout)
+		defer cn()
+
+		attempt := 1
+		backExp := backoff.NewExpBackoff(ctx.Done(), d.minBackoff, d.maxBackoff)
+		start := time.Now()
+
+		for {
+			err := dispatch(uc)
+			if err != nil {
+				if backExp.Wait() {
+					d.log.Debugf("%v action dispatch to %v with backoff attempt: %v, after %v since start", actionType, uc.component.ID, attempt, time.Since(start))
```

Comment:
Since the proxied actions types are unenroll and upgrade, and those are not frequent, it is worth bumping these logs up to the info level to make it more likely we get them by default. We can then turn them off if we find them too verbose, but I would rather start with too much logging than not enough.